Add new torus node

DolphinDream opened the pull request:
Add a Torus node to the set of generator nodes.

Note: currently the node uses only single values from the input sockets.

![torusnodetest1](https://cloud.githubusercontent.com/assets/2083719/22665163/b94a5500-ec81-11e6-9d49-eb6ece4696cf.png)

DolphinDream commented:
The Torus node has been vector-eyezed :) 

![torusnode-example2](https://cloud.githubusercontent.com/assets/2083719/22721291/c3d326ee-ed7d-11e6-82b4-de4719ea6e57.png)

DolphinDream commented:
The Torus node has been vector-eyezed :)
Here’s a dozen donuts freshly baked ;)

![torusnode-example2](https://cloud.githubusercontent.com/assets/2083719/22721897/65a5d9a0-ed81-11e6-9f31-cb3d5f635ea4.png)

DolphinDream commented:
The Torus node has been vector-eyezed :)
Here’s a dozen donuts freshly baked ;)

![torusnode-example2](https://cloud.githubusercontent.com/assets/2083719/22721961/c96bfc30-ed81-11e6-9e3f-461f773b6996.png)

ly29 commented on nodes/generator/torus.py:
```diff
@@ -0,0 +1,326 @@
+        self.outputs.new('VerticesSocket', "Normals")
+
+
+    def draw_buttons_ext(self, context, layout):
```
don't override this unless needed.

ly29 commented on nodes/generator/torus.py:
```diff
@@ -0,0 +1,326 @@
+        layout.prop(self, 'mode', expand=True)
+
+
+    def process(self):
```
```if not any(s.is_linked for s in self.outputs):```

ly29 commented on nodes/generator/torus.py:
```diff
@@ -0,0 +1,326 @@
+from bpy.props import IntProperty, FloatProperty, BoolProperty, EnumProperty
+
+from math import sin, cos, pi, sqrt, radians
+from mathutils import *
```
as far I can se you don't even user Vector, Matrix etc.
if you do please import explictly

ly29 commented on nodes/generator/torus.py:
```diff
@@ -0,0 +1,326 @@
+import time
+
+from sverchok.node_tree import SverchCustomTreeNode
+from sverchok.data_structure import updateNode, match_long_repeat, SvSetSocketAnyType
```
drop ```SvSetSocketAnyType```

ly29 commented:
Some small nitpicks but really nice work!

DolphinDream commented:
TorusKnot node has been vector-eyezed ! :)

I also addressed the latest comments and cleaned up the code.

And here's the torus knot variety pack dozen, freshly baked :) 

![torusknotvectorized-example1](https://cloud.githubusercontent.com/assets/2083719/22742024/4fea9bf2-ede3-11e6-98b7-4cc38357cd6e.png)

DolphinDream commented:
As far as I'm concerned, if there are no other issues to be addressed, the request for TorusKnot node can be closed.

ly29 commented:
Same here, I don't see any issues, just a quick test and merge

portnov commented:
Hm. For some reason the torus disappears if I toggle "separate" checkbox.

portnov commented:
Also, the torus seem to be created with inverted normals in my case.

DolphinDream commented:
I think the separate option works the same as the other generators (e.g. sphere, plane, cylinder). I believe this feature is used in conjunction with the UV connect to allow to create various loops.

![torusnode-separate-example](https://cloud.githubusercontent.com/assets/2083719/22747215/6ae1e936-edf4-11e6-950a-5ec0ade0804f.png)

portnov commented:
Ah, got it with "separate". I'm not sure why it works so, but it does not output polygons if "separate" option is on, just vertices. And yes, other generators work the same, so i think this is kind of designed to be this way.

What about normals?

DolphinDream commented:
I'm not sure what you mean by the normals being inverted. They should point
out away from the torus tube.  Can you give me an example ?

On Wed, Feb 8, 2017 at 11:57 AM, Ilya V. Portnov <notifications@github.com>
wrote:

> Ah, got it with "separate". I'm not sure why it works so, but it does not
> output polygons if "separate" option is on, just vertices. And yes, other
> generators work the same, so i think this is kind of designed to be this
> way.
>
> What about normals?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nortikin/sverchok/pull/1207#issuecomment-278388880>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB_Lh-vfNJOVXiaDZ3wq7un-fxrV55xvks5rafQUgaJpZM4L4r1A>
> .
>



-- 
--
Marius Giurgi

DolphinDream commented:
Nevermind about the normals example. I got it. I will fix it.

On Wed, Feb 8, 2017 at 11:57 AM, Ilya V. Portnov <notifications@github.com>
wrote:

> Ah, got it with "separate". I'm not sure why it works so, but it does not
> output polygons if "separate" option is on, just vertices. And yes, other
> generators work the same, so i think this is kind of designed to be this
> way.
>
> What about normals?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nortikin/sverchok/pull/1207#issuecomment-278388880>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB_Lh-vfNJOVXiaDZ3wq7un-fxrV55xvks5rafQUgaJpZM4L4r1A>
> .
>



-- 
--
Marius Giurgi

portnov commented:
ok :)

DolphinDream commented:
Okydoky. See if the latest update fixes it for you. It seems to work now at
my end.

On Wed, Feb 8, 2017 at 11:57 AM, Ilya V. Portnov <notifications@github.com>
wrote:

> Ah, got it with "separate". I'm not sure why it works so, but it does not
> output polygons if "separate" option is on, just vertices. And yes, other
> generators work the same, so i think this is kind of designed to be this
> way.
>
> What about normals?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nortikin/sverchok/pull/1207#issuecomment-278388880>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB_Lh-vfNJOVXiaDZ3wq7un-fxrV55xvks5rafQUgaJpZM4L4r1A>
> .
>



-- 
--
Marius Giurgi

portnov commented:
Yes, this works for me too now.

portnov commented:
![torus_knot_1](https://cloud.githubusercontent.com/assets/284644/22753813/a845b5aa-ee67-11e6-97a8-a0ef340cc41e.png)
